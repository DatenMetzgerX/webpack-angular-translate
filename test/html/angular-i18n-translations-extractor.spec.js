import angularI18nTranslationsExtractor from "../../src/html/angular-i18n-translations-extractor";
import StatefulHtmlParser from "../../src/html/translate-html-parser";
import Translation from "../../src/translation";

require("../translate-jest-matchers");

describe("StatefulHtmlParserSpecs", function() {
  "use strict";

  let loaderContext;

  beforeEach(function() {
    loaderContext = {
      registerTranslation: jest.fn(),
      emitError: jest.fn(),
      emitWarning: jest.fn(),
      resourcePath: "path/test.html",
      context: "path"
    };
  });

  describe("<any i18n>", function() {
    it("uses the value of the i18n attribute without the '@@' as translation id and the content as default translation", function() {
      parse("<div i18n='@@Simple Case id'>Simple case</div>");

      expect(loaderContext.registerTranslation).toHaveBeenLastCalledWith(
        new Translation("Simple Case id", "Simple case", {
          resource: "test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("emits an error if the attribute value does not contain the id indicator '@@'.", function() {
      parse("<div i18n='Simple Case id'>Simple case</div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the attribute value contains an empty id", function() {
      parse("<div i18n='@@'>Simple case</div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if no default translation is provided", function() {
      parse("<div i18n='@@Simple Case id'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the element contains multiple child nodes", function() {
      parse(
        "<div i18n='@@Simple Case id'>Created by <b>Thomas Muster</b> at 13:34</div>"
      );

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the content of the element is an expression", function() {
      parse("<div i18n='@@Simple Case id'>{{someValue}}</div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the value of the translate id is an expression", function() {
      parse("<div i18n='@@{{id}}'>Not an expression</div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if the translation element is an expression and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div suppress-dynamic-translation-error> i18n='@@Simple Case id'>{{someValue}}</div>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if the default text is an expression and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div i18n='@@{{id}}' suppress-dynamic-translation-error>Not an expression</div>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  describe("<any i18n-*>", function() {
    it("uses the value of an i18n-[attr] without '@@' as translation id and the value of the [attr] as default text", function() {
      parse(
        "<div i18n-title='@@test attribute id' title='test attribute'></div>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test attribute id", "test attribute", {
          resource: "test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("emits an error if the attribute value does not contain the id indicator '@@'.", function() {
      parse(
        "<div i18n-title='test attribute id' title='test attribute'></div>"
      );

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the attribute value contains an empty id", function() {
      parse("<div i18n-title='@@' title='test attribute'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if no corresponding attribute [attr] exist for the i18n-[attr] attribute", function() {
      parse("<div i18n-title='@@test attribute id'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the corresponding attribute [attr] is empty", function() {
      parse("<div i18n-title='@@test attribute id' title=''></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the default translation is an expression", function() {
      parse("<div  i18n-title='@@myTitle' title='{{myTitle}}'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if the i18n-[attr] uses an expression as id", function() {
      parse("<div i18n-title='@@{{id}}' title='My title'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if the i18n-[attr] id is used and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div i18n-title='@@{{id}}' title='My title' suppress-dynamic-translation-error></div>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if the default text is an expression and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div  i18n-title='@@myTitle' title='{{myTitle}}' suppress-dynamic-translation-error></div>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  function parse(source) {
    let statefulParser = new StatefulHtmlParser(loaderContext, [
      angularI18nTranslationsExtractor
    ]);
    statefulParser.parse(source);
    return statefulParser;
  }
});
