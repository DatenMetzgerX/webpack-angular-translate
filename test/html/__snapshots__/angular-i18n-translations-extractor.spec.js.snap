// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if no default translation is provided 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element <div i18n='@@Simple Case id'>...</div> with attribute  i18n is empty and is therefore missing the default translation.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if the attribute value contains an empty id 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The attribute i18n on element <div i18n='@@'>Simple case</div> defines an empty ID.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if the attribute value does not contain the id indicator '@@'. 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The attribute i18n on element <div i18n='Simple Case id'>Simple case</div> attribute is missing the custom id indicator '@@'.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if the content of the element is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div i18n='@@Simple Case id'>{{someValue}}</div>' uses an angular expression as translation id ('Simple Case id') or as default text ('{{someValue}}'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if the element contains multiple child nodes 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element <div i18n='@@Simple Case id'>Created by  at 13:34</div> has multiple child elements and, therefore, the default translation cannot be extracted.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n> emits an error if the value of the translate id is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div i18n='@@{{id}}'>Not an expression</div>' uses an angular expression as translation id ('{{id}}') or as default text ('Not an expression'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if no corresponding attribute [attr] exist for the i18n-[attr] attribute 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element <div i18n-title='@@test attribute id'>...</div> with i18n-title is missing a corresponding title attribute.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if the attribute value contains an empty id 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The attribute i18n-title on element <div i18n-title='@@' title='test attribute'>...</div> defines an empty ID.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if the attribute value does not contain the id indicator '@@'. 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The attribute i18n-title on element <div i18n-title='test attribute id' title='test attribute'>...</div> attribute is missing the custom id indicator '@@'.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if the corresponding attribute [attr] is empty 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element <div i18n-title='@@test attribute id' title=''>...</div> with i18n-title is missing a value for the corresponding title attribute.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if the default translation is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div i18n-title='@@myTitle' title='{{myTitle}}'>...</div>' uses an angular expression as translation id ('myTitle') or as default text ('{{myTitle}}'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any i18n-*> emits an error if the i18n-[attr] uses an expression as id 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div i18n-title='@@{{id}}' title='My title'>...</div>' uses an angular expression as translation id ('{{id}}') or as default text ('My title'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;
